### ISSUE TITLE
How to implement ALE plots with scaled data?

### ISSUE THREAD

**Issue by JPK01** [DanaJomar/PyALE#13]
I would like to use the ALE plot for a neural network which uses scaled input data. If I use the scaled data and the neural network in the ALE function the graph's x-axis is scaled as well. I would like the plot to not be scaled. How can I achieve this? Thanks in advance.

**Comment by DanaJomar** [comment]
Hi there!
the package does not  have out-of-the-box solution for this issue, however to address such problems, there's always the possibility to give your own matplotlib figure and axis to the function and then manipulating them as to fit your purposes.
What you need is a mapping between the original values and the scaled values and then you could use these to alter the labels shown on your plot.
I could be more helpful if you could provide me with  a reproducible example to test a few things with, but maybe this example could help you, note that this is one way to address such an issue with matplotlib and I'm sure there are multiple other options for it

``` python
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestRegressor
from PyALE import ale

def scale_feat(raw):
  max_raw = max(raw)
  min_raw = min(raw)
  scaled = (raw - min_raw)/(max_raw - min_raw)
  return scaled

def unscale_feat(scaled, max_raw, min_raw):
  raw = scaled * (max_raw - min_raw) + min_raw
  return raw

## get the raw diamond data (from R's ggplot2)
dat_diamonds = pd.read_csv(
  "https://raw.githubusercontent.com/tidyverse/ggplot2/master/data-raw/diamonds.csv"
)
X_raw = dat_diamonds.loc[:, ~dat_diamonds.columns.str.contains("price")].copy()
y = dat_diamonds["price"].copy()

features = ['carat', 'depth', 'table', 'x', 'y', 'z']
X = X_raw.copy()[features]

## scale a feature
X["table_scaled"] = scale_feat(X["table"])

## update features
features.pop(features.index("table"))
features.append("table_scaled")

## train model
model = RandomForestRegressor(random_state=1345)
model.fit(X[features], y)

## get ALE plot
fig, ax = plt.subplots()
ale_eff = ale(
    X=X[features], model=model, feature=["table_scaled"], grid_size=20, include_CI=False, fig=fig, ax=ax
)
# use `set_major_formatter` to update ticks' labels
ax.xaxis.set_major_formatter(major_formatter)
plt.show()
```

**Comment by JPK01** [comment]
Hi, thanks a lot for your helpful response. 
I tried switching the x-axis of the ALE plot by myself. I attached the code below.
As you see I scale the data with MinMaxScaler and also inverse that scaling. Then I create two ALE dataframes, one from the scaled feature data and one with the unscaled/ Inverse feature data. When I create an ALE plot in matplotlib I show the unscaled feature data in the x-asis but show the ALE effect that was computed using the scaled feature data on the y-axis. Does this make sense to you?

```
X_train_2022 = train_2022

from sklearn.preprocessing import MinMaxScaler
scaler_2022 = MinMaxScaler(feature_range=(-1,1))
X_train_2022 = pd.DataFrame(scaler_2022.fit_transform(X_train_2022), index=X_train_2022.index, columns=X_train_2022.columns)

X_train_2022_inverse = scaler_2022.inverse_transform(X_train_2022)

# Convert to DataFrame
X_train_2022_inverse_df = pd.DataFrame(X_train_2022_inverse, columns=X_train_2022.columns)

#Create ALE data from scaled data
ale_eff_scaled = ale(
    X=X_train_2022, model=model2022_1_loaded, feature=["feature"], grid_size=250, include_CI=False
)

#Create ALE data from unscaled data
ale_eff_inverse = ale(
    X=X_train_2022_inverse_df, model=model2022_1_loaded, feature=["feature"], grid_size=250, include_CI=False
)

#Replace scaled "feature" column with unscaled "feature" column
ale_eff_scaled.reset_index(inplace=True)
ale_eff_inverse.reset_index(inplace=True)

ale_eff_scaled.update(ale_eff_inverse['feature'])


# Create data
x = ale_eff_scaled['feature']
y = ale_eff_scaled['eff']

# Create a new figure and axes with a specified size and DPI
fig, ax = plt.subplots(figsize=(12, 8), dpi=150)

# Plot your line graph
ax.plot(x, y, label='Line Plot', linewidth=2.5)

# Reduce the size of the tick labels
ax.tick_params(axis='both', which='major', labelsize=20)

# Add a rugplot at the bottom.
sns.rugplot(data=train_2022["feature"], ax=ax, height=0.05, axis='x', color='black')

# Set the limit of y-axis slightly below the minimum y to include the rug plot 
y_range = max(y) - min(y)
ax.set_ylim([min(y) - 0.1 * y_range, max(y)])  # leave 10% of the y-range as space at the bottom
#ax.set_ylim([min(y), max(y)]) 

buffer = 0.08 * (y.max() - y.min())  # 8% buffer
ax.set_ylim(y.min() - buffer, y.max() + buffer)

plt.xlabel('Feature', fontsize= 20)
plt.ylabel('Effect on prediction (centered)', fontsize= 20)
plt.show()
```

**Comment by DanaJomar** [comment]
An example workaround has been added to
https://github.com/DanaJomar/PyALE/blob/master/examples/

See the notebook `PyALE with Scaled Features.ipynb` or the rendered html version `PyALE with Scaled Features.html`
